source/common/config/delta_subscription_state.cc

```diff
@@ -160,19 +164,32 @@ DeltaSubscriptionState::getNextRequestWithAck(const UpdateAck& ack) {
   return request;
 }
 
-void DeltaSubscriptionState::setResourceVersion(const std::string& resource_name,
-                                                const std::string& resource_version) {
-  resource_versions_[resource_name] = ResourceVersion(resource_version);
-  resource_names_.insert(resource_name);
+void DeltaSubscriptionState::addResourceState(
+    const envoy::service::discovery::v3::Resource& resource) {
+  Event::TimerPtr ttl_timer;
+  if (resource.has_ttl()) {
+    ttl_timer = dispatcher_.createTimer([this, resource]() -> void {
```

Comment:
What are the performance implications of creating around O(10k) timers and having them fire at the same time? We've seen some resource types (e.g. EDS) contain this number of resources. I'd just like to make sure that nobody is surprised when we turn this on day one. Possibly might motivate a runtime flag.

Reply:
I'm not sure what the overhead involved would be, maybe @antoniovicente has an idea? 

I'm fine with adding a runtime flag, though this would have to be explicitly configured by the server to do anything. Or are you concerned with *aaS services that might want to enable this and end users being overwhelmed?

Reply:
I think normally a delta response that removes all the resources would do so in one go (`removed_resources` = full list of resources), but at the same time I don't see EDS handling this in batch. In fact, it seems like EDS always operates on one resource at a time and doesn't even seem to respect `removed_resources`, so maybe this wouldn't even work with EDS in its current state? As written I don't see how EDS in particular would benefit from batching/jittering the callbacks, though I can imagine it generally being a good idea to be careful with perf here. 

Do we feel like we need to optimize this now or is a runtime flag around this fine for now sufficient?

Reply:
I think I'ld like to understand typical usage better before going to far into optimization.  @htuch thoughts?

Reply:
The main point is that if we have one timer per resource, and we have O(10k) resources, we can get into some bad situations. But, since this is opt-in and won't affect non-TTL users, I think it's fine to go with what is here for now. Just wanted to raise this as a limitation and potential TODO consideration.